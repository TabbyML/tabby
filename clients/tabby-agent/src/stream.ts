import { Readable } from "readable-stream";
import { readableFromWeb } from "readable-from-web";
import { EventSourceParserStream, ParsedEvent } from "eventsource-parser/stream";
import type { components as TabbyApiComponents } from "tabby-openapi/compatible";
import { getLogger } from "./logger";

const logger = getLogger("StreamParser");

export function readChatStream(stream: ReadableStream, signal?: AbortSignal): Readable {
  const eventStream = stream.pipeThrough(new TextDecoderStream()).pipeThrough(new EventSourceParserStream());
  const readableStream = readableFromWeb(eventStream, { objectMode: true });
  return readableStream.map(
    (event: ParsedEvent): string | undefined => {
      try {
        if (event.type === "event") {
          const chunk = JSON.parse(event.data) as TabbyApiComponents["schemas"]["ChatCompletionChunk"];
          const text = chunk.choices[0]?.delta.content;
          if (typeof text === "string") {
            return text;
          }
        }
      } catch (error) {
        logger.error("Failed to parse chat stream chunk.", error);
        logger.trace("Parsing failed with event:", { event });
      }
      return undefined;
    },
    { signal },
  );
}

export async function parseChatResponse(response: Response, signal?: AbortSignal): Promise<string> {
  if (!response.body) {
    return "";
  }
  const readableStream = readChatStream(response.body, signal);
  return readableStream.reduce<string>((text, delta) => text + delta, "", { signal });
}
