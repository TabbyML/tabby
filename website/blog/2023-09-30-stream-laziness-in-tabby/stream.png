version https://git-lfs.github.com/spec/v1
oid sha256:74734640c70fb1ebbfe6baa9a3ff8f8ef257577766d98df6bbdb066e44a9a17d
size 104103
