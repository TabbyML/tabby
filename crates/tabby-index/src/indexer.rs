use std::collections::HashSet;

use anyhow::{bail, Result};
use async_stream::stream;
use futures::{stream::BoxStream, Stream, StreamExt};
use serde_json::json;
use tabby_common::{
    index::{IndexSchema, FIELD_SOURCE_ID},
    path,
};
use tantivy::{
    aggregation::{
        agg_req::Aggregation,
        agg_result::{AggregationResult, BucketResult},
        AggregationCollector, Key,
    },
    collector::TopDocs,
    doc,
    query::AllQuery,
    schema::{self, Value},
    DocAddress, DocSet, IndexWriter, Searcher, TantivyDocument, Term, TERMINATED,
};
use tokio::{sync::mpsc, task::JoinHandle};
use tracing::{debug, warn};

use crate::tantivy_utils::open_or_create_index;

pub struct IndexId {
    pub source_id: String,
    pub id: String,
}

pub trait ToIndexId {
    fn to_index_id(&self) -> IndexId;
}

#[async_trait::async_trait]
pub trait IndexAttributeBuilder<T>: Send + Sync {
    /// Build document level attributes, these attributes are only stored but not indexed.
    async fn build_attributes(&self, document: &T) -> serde_json::Value;

    /// Build chunk level attributes, these attributes are stored and indexed.
    async fn build_chunk_attributes<'a>(
        &self,
        document: &'a T,
    ) -> BoxStream<'a, JoinHandle<Result<(Vec<String>, serde_json::Value)>>>;
}

pub struct TantivyDocBuilder<T> {
    corpus: &'static str,
    builder: Box<dyn IndexAttributeBuilder<T>>,
}

impl<T: ToIndexId> TantivyDocBuilder<T> {
    pub fn new(corpus: &'static str, builder: impl IndexAttributeBuilder<T> + 'static) -> Self {
        Self {
            corpus,
            builder: Box::new(builder),
        }
    }

    pub async fn build(
        &self,
        document: T,
    ) -> (
        String,
        impl Stream<Item = JoinHandle<Option<TantivyDocument>>> + '_,
    ) {
        let schema = IndexSchema::instance();
        let IndexId { source_id, id } = document.to_index_id();

        let now = tantivy::time::OffsetDateTime::now_utc();
        let updated_at = tantivy::DateTime::from_utc(now);

        let cloned_id = id.clone();
        let doc_id = id.clone();
        let doc_attributes = self.builder.build_attributes(&document).await;
        let s = stream! {
            let (tx, mut rx) = mpsc::channel(32);

            for await chunk_doc in self.build_chunks(cloned_id, source_id.clone(), updated_at, document).await {
                let tx = tx.clone();
                let doc_id = doc_id.clone();
                yield tokio::spawn(async move {
                    match chunk_doc.await {
                        Ok(Ok(doc)) => {
                            Some(doc)
                        }
                        Ok(Err(e)) => {
                            warn!("Failed to build chunk for document '{}': {}", doc_id, e);
                            tx.send(()).await.unwrap_or_else(|e| {
                                warn!("Failed to send error signal for document '{}': {}", doc_id, e);
                            });
                            None
                        }
                        Err(e) => {
                            warn!("Failed to call build chunk '{}': {}", doc_id, e);
                            tx.send(()).await.unwrap_or_else(|e| {
                                warn!("Failed to send error signal for document '{}': {}", doc_id, e);
                            });
                            None
                        }
                    }
                });
            };

            // drop tx to signal the end of the stream
            // the cloned is dropped in its own thread
            drop(tx);

            let mut doc = doc! {
                schema.field_id => doc_id,
                schema.field_source_id => source_id,
                schema.field_corpus => self.corpus,
                schema.field_attributes => doc_attributes,
                schema.field_updated_at => updated_at,
            };

            yield tokio::spawn(async move {
                let mut failed_count = 0;
                while let Some(_) = rx.recv().await {
                    failed_count += 1;
                }
                if failed_count > 0 {
                    doc.add_u64(schema.field_failed_chunks_count, failed_count as u64);
                }
                Some(doc)
             });
        };

        (id, s)
    }

    async fn build_chunks(
        &self,
        id: String,
        source_id: String,
        updated_at: tantivy::DateTime,
        document: T,
    ) -> impl Stream<Item = JoinHandle<Result<TantivyDocument>>> + '_ {
        let kind = self.corpus;
        stream! {
            let schema = IndexSchema::instance();
            for await (chunk_id, task) in self.builder.build_chunk_attributes(&document).await.enumerate() {
                let id = id.clone();
                let source_id = source_id.clone();

                yield tokio::spawn(async move {
                    let built_chunk_attributes_result = task.await?;
                    let (tokens, chunk_attributes) = built_chunk_attributes_result?;

                    let mut doc = doc! {
                        schema.field_id => id,
                        schema.field_source_id => source_id,
                        schema.field_corpus => kind,
                        schema.field_updated_at => updated_at,
                        schema.field_chunk_id => format!("{}-{}", id, chunk_id),
                        schema.field_chunk_attributes => chunk_attributes,
                    };

                    for token in &tokens {
                        doc.add_text(schema.field_chunk_tokens, token);
                    }

                    Ok(doc)
                });
            }
        }
    }

    pub async fn backfill_doc_attributes(
        &self,
        origin: &TantivyDocument,
        doc: &T,
    ) -> TantivyDocument {
        let schema = IndexSchema::instance();
        let mut doc = doc! {
            schema.field_id => get_text(origin, schema.field_id),
            schema.field_source_id => get_text(origin, schema.field_source_id).to_string(),
            schema.field_corpus => get_text(origin, schema.field_corpus).to_string(),
            schema.field_attributes => self.builder.build_attributes(doc).await,
            schema.field_updated_at => get_date(origin, schema.field_updated_at),
        };
        if let Some(failed_chunks) = get_number_optional(origin, schema.field_failed_chunks_count) {
            doc.add_u64(schema.field_failed_chunks_count, failed_chunks as u64);
        }

        doc
    }
}

pub struct Indexer {
    corpus: String,
    searcher: Searcher,
    writer: IndexWriter,
}

impl Indexer {
    pub fn new(corpus: &str) -> Self {
        let doc = IndexSchema::instance();
        let (_, index) = open_or_create_index(&doc.schema, &path::index_dir());
        let writer = index
            .writer(150_000_000)
            .expect("Failed to create index writer");
        let reader = index.reader().expect("Failed to create index reader");

        Self {
            corpus: corpus.to_owned(),
            searcher: reader.searcher(),
            writer,
        }
    }

    pub async fn add(&self, document: TantivyDocument) {
        self.writer
            .add_document(document)
            .expect("Failed to add document");
    }

    pub async fn get_doc(&self, id: &str) -> Result<TantivyDocument> {
        let schema = IndexSchema::instance();
        let query = schema.doc_query(&self.corpus, id);
        let docs = match self.searcher.search(&query, &TopDocs::with_limit(1)) {
            Ok(docs) => docs,
            Err(e) => {
                debug!("query tantivy error: {}", e);
                return Err(e.into());
            }
        };
        if docs.is_empty() {
            bail!("Document not found: {}", id);
        }

        self.searcher
            .doc(docs.first().unwrap().1)
            .map_err(|e| e.into())
    }

    pub fn delete(&self, id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.doc_query_with_chunks(&self.corpus, id)));
    }

    pub fn delete_doc(&self, id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.doc_query(&self.corpus, id)));
    }

    pub fn commit(mut self) {
        self.writer.commit().expect("Failed to commit changes");
        self.writer
            .wait_merging_threads()
            .expect("Failed to wait for merging threads");
    }

    // Check whether the document ID presents in the corpus.
    pub fn is_indexed(&self, id: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_query(&self.corpus, id);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };
        !docs.is_empty()
    }

    /// Iterates over all the document IDs in the corpus.
    pub fn iter_ids(&self) -> impl Stream<Item = String> + '_ {
        let schema = IndexSchema::instance();

        stream! {
            // Based on https://github.com/quickwit-oss/tantivy/blob/main/examples/iterating_docs_and_positions.rs
            for (segment_ordinal, segment_reader) in self.searcher.segment_readers().iter().enumerate() {
                let Ok(inverted_index) = segment_reader.inverted_index(schema.field_corpus) else {
                    continue;
                };

                let term_corpus = Term::from_field_text(schema.field_corpus, &self.corpus);
                let Ok(Some(mut postings)) = inverted_index.read_postings(&term_corpus, tantivy::schema::IndexRecordOption::Basic) else {
                    continue;
                };

                let mut doc_id = postings.doc();
                while doc_id != TERMINATED {
                    if !segment_reader.is_deleted(doc_id) {
                        let doc_address = DocAddress::new(segment_ordinal as u32, doc_id);
                        let doc: TantivyDocument = self.searcher.doc(doc_address).expect("Failed to read document");

                        // Skip chunks, as we only want to iterate over the main docs
                        if doc.get_first(schema.field_chunk_id).is_none() {
                            let id = get_text(&doc, schema.field_id);
                            yield id.to_owned();
                        }
                    }
                    doc_id = postings.advance();
                }
            }
        }
    }

    pub fn is_indexed_after(&self, id: &str, time: chrono::DateTime<chrono::Utc>) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_indexed_after(&self.corpus, id, time);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };

        !docs.is_empty()
    }

    /// Check whether the document has failed chunks.
    ///
    /// failed chunks tracks the number of embedding indexing failed chunks for a document.
    pub fn has_failed_chunks(&self, id: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_has_failed_chunks(&self.corpus, id);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };

        !docs.is_empty()
    }

    // Check whether the document has attribute field.
    pub fn has_attribute_field(&self, id: &str, field: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_has_attribute_field(&self.corpus, id, field);
        match self.searcher.search(&query, &TopDocs::with_limit(1)) {
            Ok(docs) => !docs.is_empty(),
            Err(e) => {
                debug!("query tantivy error: {}", e);
                false
            }
        }
    }
}

pub struct IndexGarbageCollector {
    searcher: Searcher,
    writer: IndexWriter,
}

impl IndexGarbageCollector {
    pub fn new() -> Self {
        let doc = IndexSchema::instance();
        let (_, index) = open_or_create_index(&doc.schema, &path::index_dir());
        let writer = index
            .writer(150_000_000)
            .expect("Failed to create index writer");
        let reader = index.reader().expect("Failed to create index reader");

        Self {
            searcher: reader.searcher(),
            writer,
        }
    }

    pub fn garbage_collect(&self, active_source_ids: &[String]) -> anyhow::Result<()> {
        let source_ids: HashSet<_> = active_source_ids.iter().collect();

        let count_aggregation: Aggregation = serde_json::from_value(json!({
            "terms": {
                "field": FIELD_SOURCE_ID,
            }
        }))
        .unwrap();

        let collector = AggregationCollector::from_aggs(
            vec![("count".to_owned(), count_aggregation)]
                .into_iter()
                .collect(),
            Default::default(),
        );

        let res = self.searcher.search(&AllQuery, &collector)?;
        let Some(AggregationResult::BucketResult(BucketResult::Terms { buckets, .. })) =
            res.0.get("count")
        else {
            bail!("Failed to get source_id count");
        };
        for source_id in buckets {
            let count = source_id.doc_count;
            let Key::Str(source_id) = &source_id.key else {
                warn!("Failed to get source_id key as string");
                continue;
            };

            if !source_ids.contains(source_id) {
                debug!("Deleting {} documents for source_id: {}", count, source_id,);
                self.delete_by_source_id(source_id);
            }
        }

        Ok(())
    }

    fn delete_by_source_id(&self, source_id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.source_id_query(source_id)));
    }

    pub fn commit(mut self) {
        self.writer.commit().expect("Failed to commit changes");
        self.writer
            .wait_merging_threads()
            .expect("Failed to wait for merging threads");
    }
}

fn get_text(doc: &TantivyDocument, field: schema::Field) -> &str {
    doc.get_first(field).unwrap().as_str().unwrap()
}

fn get_date(doc: &TantivyDocument, field: schema::Field) -> tantivy::DateTime {
    doc.get_first(field).unwrap().as_datetime().unwrap()
}

fn get_number_optional(doc: &TantivyDocument, field: schema::Field) -> Option<i64> {
    doc.get_first(field)?.as_i64()
}
