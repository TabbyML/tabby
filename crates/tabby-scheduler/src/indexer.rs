use std::collections::HashSet;

use anyhow::bail;
use async_stream::stream;
use futures::{stream::BoxStream, Stream, StreamExt};
use serde_json::json;
use tabby_common::{index::IndexSchema, path};
use tantivy::{
    aggregation::{
        agg_req::Aggregation,
        agg_result::{AggregationResult, BucketResult},
        AggregationCollector, Key,
    },
    collector::TopDocs,
    doc,
    schema::{self, Value},
    DocAddress, DocSet, IndexWriter, Searcher, TantivyDocument, Term, TERMINATED,
};
use tokio::task::JoinHandle;
use tracing::{debug, warn};

use crate::tantivy_utils::open_or_create_index;

pub struct IndexId {
    pub source_id: String,
    pub id: String,
}

pub trait ToIndexId {
    fn to_index_id(&self) -> IndexId;
}

#[async_trait::async_trait]
pub trait IndexAttributeBuilder<T>: Send + Sync {
    async fn build_attributes(&self, document: &T) -> serde_json::Value;
    async fn build_chunk_attributes(
        &self,
        document: &T,
    ) -> BoxStream<JoinHandle<(Vec<String>, serde_json::Value)>>;
}

pub struct TantivyDocBuilder<T> {
    corpus: &'static str,
    builder: Box<dyn IndexAttributeBuilder<T>>,
}

impl<T: ToIndexId> TantivyDocBuilder<T> {
    pub fn new(corpus: &'static str, builder: impl IndexAttributeBuilder<T> + 'static) -> Self {
        Self {
            corpus,
            builder: Box::new(builder),
        }
    }

    pub async fn build(
        &self,
        document: T,
    ) -> (
        String,
        impl Stream<Item = JoinHandle<Option<TantivyDocument>>> + '_,
    ) {
        let schema = IndexSchema::instance();
        let IndexId { source_id, id } = document.to_index_id();

        let now = tantivy::time::OffsetDateTime::now_utc();
        let updated_at = tantivy::DateTime::from_utc(now);

        let doc = doc! {
            schema.field_id => id,
            schema.field_source_id => source_id,
            schema.field_corpus => self.corpus,
            schema.field_attributes => self.builder.build_attributes(&document).await,
            schema.field_updated_at => updated_at,
        };

        let cloned_id = id.clone();
        let s = stream! {
            yield tokio::spawn(async move { Some(doc) });

            for await doc in self.build_chunks(cloned_id, source_id, updated_at, document).await {
                yield doc;
            }
        };

        (id, s)
    }

    async fn build_chunks(
        &self,
        id: String,
        source_id: String,
        updated_at: tantivy::DateTime,
        document: T,
    ) -> impl Stream<Item = JoinHandle<Option<TantivyDocument>>> + '_ {
        let kind = self.corpus;
        stream! {
            let schema = IndexSchema::instance();
            for await (chunk_id, task) in self.builder.build_chunk_attributes(&document).await.enumerate() {
                let id = id.clone();
                let source_id = source_id.clone();

                yield tokio::spawn(async move {
                    let Ok((tokens, chunk_attributes)) = task.await else {
                        return None;
                    };

                    let mut doc = doc! {
                        schema.field_id => id,
                        schema.field_source_id => source_id,
                        schema.field_corpus => kind,
                        schema.field_updated_at => updated_at,
                        schema.field_chunk_id => format!("{}-{}", id, chunk_id),
                        schema.field_chunk_attributes => chunk_attributes,
                    };

                    for token in tokens {
                        doc.add_text(schema.field_chunk_tokens, token);
                    }

                    Some(doc)
                });
            }
        }
    }
}

pub struct Indexer {
    corpus: String,
    searcher: Searcher,
    writer: IndexWriter,
    pub recreated: bool,
}

impl Indexer {
    pub fn new(corpus: &str) -> Self {
        let doc = IndexSchema::instance();
        let (recreated, index) = open_or_create_index(&doc.schema, &path::index_dir());
        let writer = index
            .writer(150_000_000)
            .expect("Failed to create index writer");
        let reader = index.reader().expect("Failed to create index reader");

        Self {
            corpus: corpus.to_owned(),
            searcher: reader.searcher(),
            writer,
            recreated,
        }
    }

    pub async fn add(&self, document: TantivyDocument) {
        self.writer
            .add_document(document)
            .expect("Failed to add document");
    }

    pub fn delete(&self, id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.doc_query_with_chunks(&self.corpus, id)));
    }

    fn delete_by_source_id(&self, source_id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.source_query(&self.corpus, source_id)));
    }

    pub fn commit(mut self) {
        self.writer.commit().expect("Failed to commit changes");
        self.writer
            .wait_merging_threads()
            .expect("Failed to wait for merging threads");
    }

    // Check whether the document ID presents in the corpus.
    pub fn is_indexed(&self, id: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_query(&self.corpus, id);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };
        !docs.is_empty()
    }

    /// Iterates over all the document IDs in the corpus.
    pub fn iter_ids(&self) -> impl Stream<Item = String> + '_ {
        let schema = IndexSchema::instance();

        stream! {
            // Based on https://github.com/quickwit-oss/tantivy/blob/main/examples/iterating_docs_and_positions.rs
            for (segment_ordinal, segment_reader) in self.searcher.segment_readers().iter().enumerate() {
                let Ok(inverted_index) = segment_reader.inverted_index(schema.field_corpus) else {
                    continue;
                };

                let term_corpus = Term::from_field_text(schema.field_corpus, &self.corpus);
                let Ok(Some(mut postings)) = inverted_index.read_postings(&term_corpus, tantivy::schema::IndexRecordOption::Basic) else {
                    continue;
                };

                let prefix_to_strip = format!("{}:", self.corpus);
                let mut doc_id = postings.doc();
                while doc_id != TERMINATED {
                    if !segment_reader.is_deleted(doc_id) {
                        let doc_address = DocAddress::new(segment_ordinal as u32, doc_id);
                        let doc: TantivyDocument = self.searcher.doc(doc_address).expect("Failed to read document");

                        // Skip chunks, as we only want to iterate over the main docs
                        if doc.get_first(schema.field_chunk_id).is_none() {
                            if let Some(id) = get_text(&doc, schema.field_id).strip_prefix(&prefix_to_strip) {
                                yield id.to_owned();
                            }
                        }
                    }
                    doc_id = postings.advance();
                }
            }
        }
    }

    pub fn garbage_collect(&self, active_source_ids: &[String]) -> anyhow::Result<()> {
        let source_ids: HashSet<_> = active_source_ids.iter().collect();
        let schema = IndexSchema::instance();

        let count_aggregation: Aggregation = serde_json::from_value(json!({
            "terms": {
                "field": "source_id"
            }
        }))
        .unwrap();

        let collector = AggregationCollector::from_aggs(
            vec![("count".to_owned(), count_aggregation)]
                .into_iter()
                .collect(),
            Default::default(),
        );

        let res = self
            .searcher
            .search(&schema.corpus_query(&self.corpus), &collector)?;
        let Some(AggregationResult::BucketResult(BucketResult::Terms { buckets, .. })) =
            res.0.get("count")
        else {
            bail!("Failed to get source_id count");
        };
        for source_id in buckets {
            let Key::Str(source_id) = &source_id.key else {
                warn!("Failed to get source_id key as string");
                continue;
            };

            if !source_ids.contains(source_id) {
                debug!(
                    "Deleting documents for source_id: {} in corpus: {}",
                    source_id, self.corpus
                );
                self.delete_by_source_id(source_id);
            }
        }

        Ok(())
    }
}

fn get_text(doc: &TantivyDocument, field: schema::Field) -> &str {
    doc.get_first(field).unwrap().as_str().unwrap()
}
