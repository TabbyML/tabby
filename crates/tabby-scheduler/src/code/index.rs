use std::{pin::pin, sync::Arc};

use async_stream::stream;
use futures::StreamExt;
use ignore::{DirEntry, Walk};
use tabby_common::{config::RepositoryConfig, index::corpus};
use tabby_inference::Embedding;
use tracing::warn;

use super::{
    create_code_builder,
    intelligence::{CodeIntelligence, SourceCode},
};
use crate::indexer::Indexer;

// Magic numbers
static MAX_LINE_LENGTH_THRESHOLD: usize = 300;
static AVG_LINE_LENGTH_THRESHOLD: f32 = 150f32;
static MIN_ALPHA_NUM_FRACTION: f32 = 0.25f32;
static MAX_NUMBER_OF_LINES: usize = 100000;
static MAX_NUMBER_FRACTION: f32 = 0.5f32;

pub async fn index_repository(embedding: Arc<dyn Embedding>, repository: &RepositoryConfig) {
    let total_files = Walk::new(repository.dir()).count();
    let file_stream = stream! {
        for file in Walk::new(repository.dir()) {
            let file = match file {
                Ok(file) => file,
                Err(e) => {
                    warn!("Failed to walk file tree for indexing: {e}");
                    continue;
                }
            };

            yield file;
        }
    }
    // Commit every 100 files
    .chunks(100);

    let mut file_stream = pin!(file_stream);

    let mut count_files = 0;
    let mut count_chunks = 0;
    while let Some(files) = file_stream.next().await {
        count_files += files.len();
        count_chunks += add_changed_documents(repository, embedding.clone(), files).await;
        logkit::info!(
            "{}/{} files has been processed, {} docs has been indexed",
            count_files,
            total_files,
            count_chunks
        );
    }
}

pub async fn garbage_collection() {
    let index = Indexer::new(corpus::CODE);
    stream! {
        let mut num_to_keep = 0;
        let mut num_to_delete = 0;

        for await id in index.iter_ids() {
            let item_key = id;
            if CodeIntelligence::check_source_file_id_matched(&item_key) {
                num_to_keep += 1;
            } else {
                num_to_delete += 1;
                index.delete(&item_key);
            }
        }

        logkit::info!("Finished garbage collection for code index: {num_to_keep} items kept, {num_to_delete} items removed");
        index.commit();
    }.collect::<()>().await;
}

async fn add_changed_documents(
    repository: &RepositoryConfig,
    embedding: Arc<dyn Embedding>,
    files: Vec<DirEntry>,
) -> usize {
    let concurrency = std::cmp::max(std::thread::available_parallelism().unwrap().get() * 4, 64);
    let builder = Arc::new(create_code_builder(Some(embedding)));
    let index = Arc::new(Indexer::new(corpus::CODE));
    let cloned_index = index.clone();
    let repository = repository.clone();

    let mut count_docs = 0;
    stream! {
        for file in files {
            let Some(key) = CodeIntelligence::compute_source_file_id(file.path()) else {
                continue;
            };

            if cloned_index.is_indexed(&key) {
                // Skip if already indexed
                continue;
            }

            let Some(code) = CodeIntelligence::compute_source_file(&repository, file.path()) else {
                continue;
            };

            if !is_valid_file(&code) {
                continue;
            }

            let (_, s) = builder.build(code).await;
            for await task in s {
                yield task;
            }
        }
    }
    .buffer_unordered(concurrency)
    .filter_map(|x| async { x.ok().flatten() })
    .for_each(|x| {
        count_docs += 1;
        index.add(x)
    })
    .await;

    match Arc::try_unwrap(index) {
        Ok(index) => index.commit(),
        Err(_) => {
            panic!("Failed to commit code index");
        }
    };

    count_docs
}

fn is_valid_file(file: &SourceCode) -> bool {
    file.max_line_length <= MAX_LINE_LENGTH_THRESHOLD
        && file.avg_line_length <= AVG_LINE_LENGTH_THRESHOLD
        && file.alphanum_fraction >= MIN_ALPHA_NUM_FRACTION
        && file.num_lines <= MAX_NUMBER_OF_LINES
        && file.number_fraction <= MAX_NUMBER_FRACTION
}

#[cfg(test)]
mod tests {
    use futures::StreamExt;
    use insta::assert_snapshot;

    use crate::code::intelligence::CodeIntelligence;

    #[tokio::test]
    async fn test_code_splitter() {
        // First file, chat/openai_chat.rs
        let file_contents = include_str!("../../../http-api-bindings/src/chat/openai_chat.rs");

        let rust_chunks = CodeIntelligence::chunks(file_contents, "rust")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", rust_chunks));

        let text_chunks = CodeIntelligence::chunks(file_contents, "unknown")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", text_chunks));

        // Second file, tabby-db/src/cache.rs
        let file_contents2 = include_str!("../../../../ee/tabby-db/src/cache.rs");

        let rust_chunks2 = CodeIntelligence::chunks(file_contents2, "rust")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", rust_chunks2));

        let text_chunks2 = CodeIntelligence::chunks(file_contents2, "unknown")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", text_chunks2));
    }
}
