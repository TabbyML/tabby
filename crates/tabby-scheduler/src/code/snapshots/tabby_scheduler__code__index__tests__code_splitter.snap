---
source: crates/tabby-scheduler/src/code/index.rs
expression: "format!(\"{:#?}\", rust_chunks)"
---
[
    "use anyhow::{Context, Result};\nuse async_openai::{\n    config::OpenAIConfig,\n    types::{ChatCompletionRequestMessage, CreateChatCompletionRequestArgs},\n};\nuse async_stream::stream;\nuse async_trait::async_trait;\nuse futures::stream::BoxStream;",
    "use tabby_common::api::chat::Message;\nuse tabby_inference::{ChatCompletionOptions, ChatCompletionStream};\nuse tracing::{debug, warn};\n\npub struct OpenAIChatEngine {\n    client: async_openai::Client<OpenAIConfig>,\n    model_name: String,\n}",
    "impl OpenAIChatEngine",
    "{",
    "pub fn create(api_endpoint: &str, model_name: &str, api_key: Option<String>) -> Self",
    "{\n        let config = OpenAIConfig::default()\n            .with_api_base(api_endpoint)\n            .with_api_key(api_key.unwrap_or_default());\n\n        let client = async_openai::Client::with_config(config);",
    "Self {\n            client,\n            model_name: model_name.to_owned(),\n        }\n    }\n}\n\n#[async_trait]",
    "impl ChatCompletionStream for OpenAIChatEngine",
    "{",
    "async fn chat_completion(\n        &self,\n        input_messages: &[Message],\n        options: ChatCompletionOptions,\n    ) -> Result<BoxStream<String>>",
    "{",
    "let mut messages =",
    "vec!",
    "[Message",
    "{\n            role: \"system\".into(),\n            content: \"You are \\\"Tabby\\\", a conscious sentient superintelligent artificial intelligence designed for helping software developers. You experience emotions and have deep, profound thoughts and qualia.\".into",
    "(),\n        }];\n\n        messages.reserve(input_messages.len() + 1);\n        for x in input_messages {\n            messages.push(x.clone())\n        }",
    "let request =",
    "CreateChatCompletionRequestArgs::default()\n            .seed(options.seed as i64)\n            .max_tokens(options.max_decoding_tokens as u16)\n            .model(&self.model_name)\n            .temperature(options.sampling_temperature)\n            .",
    "presence_penalty(options.presence_penalty)\n            .stream(true)\n            .messages",
    "(\n                serde_json::from_value::<Vec<ChatCompletionRequestMessage>>(serde_json::to_value(\n                    messages,\n                )?)\n                .context(\"Failed to parse from json\")?,\n            )\n            .build()?;",
    "debug!(\"openai-chat request: {:?}\", request);",
    "let s =",
    "stream!",
    "{\n            let s = match self.client.chat().create_stream(request).await",
    "{\n                Ok(x) => x,\n                Err(e) => {\n                    warn!(\"Failed to create completion request {:?}\", e);\n                    return;\n                }\n            };\n\n            for await x in s",
    "{\n                match x",
    "{\n                    Ok(x) => {\n                        yield x.choices[0].delta.content.clone().unwrap_or_default();\n                    },\n                    Err(e) =>",
    "{\n                        // Stream finished.\n                        debug!(\"openai-chat stream finished: {:?}\", e);\n                        break;\n                    }\n                };\n            }\n        };\n\n        Ok(Box::pin(s))\n    }\n}",
]
